File: AccuPay/Domain/TimeLogs/TimeLog.vb
Option Strict On

Imports System.ComponentModel.DataAnnotations
Imports System.ComponentModel.DataAnnotations.Schema

Namespace Global.AccuPay.Entity

    <Table("employeetimeentrydetails")>
    Public Class TimeLog

        <Key>
        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property RowID As Integer?

        Public Property OrganizationID As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property Created As Date

        Public Property CreatedBy As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Computed)>
        Public Property LastUpd As Date?

        Public Property LastUpdBy As Integer?

        Public Property EmployeeID As Integer?

        <Column("Date")>
        Public Property LogDate As Date

        Public Property TimeIn As TimeSpan?

        Public Property TimeOut As TimeSpan?

        Public Property TimeStampIn As Date?

        Public Property TimeStampOut As Date?

        Public Property TimeentrylogsImportID As String
        Public Property BranchID As Integer?

        Public ReadOnly Property HasLogs As Boolean
            Get
                Return TimeIn.HasValue And TimeOut.HasValue
            End Get
        End Property

        Public Sub New()
        End Sub

        Public Sub New(timeIn As String, timeOut As String)
            Me.TimeIn = TimeSpan.Parse(timeIn)
            Me.TimeOut = TimeSpan.Parse(timeOut)
        End Sub

        <ForeignKey("EmployeeID")>
        Public Overridable Property Employee As Employee

    End Class

End Namespace

File: Accupay.Data/Entities/TimeLog.cs
using AccuPay.Utilities.Extensions;
using System;
using System.ComponentModel.DataAnnotations.Schema;

namespace AccuPay.Data.Entities
{
    [Table("employeetimeentrydetails")]
    public class TimeLog : EmployeeDataEntity
    {
        [Column("Date")]
        public DateTime LogDate { get; set; }

        public TimeSpan? TimeIn { get; set; }

        public TimeSpan? TimeOut { get; set; }

        public DateTime? TimeStampIn { get; set; }

        public DateTime? TimeStampOut { get; set; }

        public string TimeentrylogsImportID { get; set; }

        public int? BranchID { get; set; }

        [ForeignKey("EmployeeID")]
        public virtual Employee Employee { get; set; }

        [ForeignKey("BranchID")]
        public virtual Branch Branch { get; set; }

        // Eventually just use TimeStampIn and TimeStampOut instead
        // of using TimeInFull and TimeOutFull. But since TimeStampIn and TimeStampOut
        // is created by Lambert, its data is not reliable.
        [NotMapped]
        public DateTime? TimeInFull
        {
            get => TimeIn == null ?
                        (DateTime?)null :
                        LogDate.Date.ToMinimumHourValue().Add(TimeIn.Value);

            set => TimeIn = value == null ? null : value?.TimeOfDay;
        }

        [NotMapped]
        public DateTime? TimeOutFull
        {
            get => TimeOut == null ?
                        (DateTime?)null :
                        LogDate.Date.ToMinimumHourValue().Add(TimeOut.Value);

            set => TimeOut = value == null ? null : value?.TimeOfDay;
        }
    }
}
